libs/apigateway-lambda-proxy/src/lib/metric-alarms.ts

```diff
@@ -0,0 +1,376 @@
+ *
+ * @param api the APIGateway to test
+ * @returns true if api.id is undefined
+ */
+function isRestApi(api: ApiGateway) {
+    return api.id === undefined
+}
+
+/**
+ * Check if an ApiGateway is an HTTP API
+ *
+ * @param api the APIGateway to test
+ * @returns true if api.id is defined
+ */
+function isHttpApi(api: ApiGateway) {
+    return !isRestApi(api)
+}
+
+/**
+ * Get dimensions for a cloudwatch alarm
+ *
+ * @param api the APIGateway to test
+ * @returns the dimensions for passing to CloudWatch alarms
+ */
+function getDimensions(api: ApiGateway) {
+    return pulumi
+        .all([api.id, api.name, api.stage])
+        .apply(([id, name, stage]) => ({
+            ...(isHttpApi(api)
+                ? {
+                      ApiId: id,
+                  }
+                : {
+                      ApiName: name,
+                  }),
+            ...(stage
+                ? {
+                      Stage: stage,
+                  }
+                : {}),
+        }))
+}
+
+/** raise an alarm if the 5xx error rate exceeds errorRate5xxPercent */
+export class ErrorRate5xxAlarm extends pulumi.ComponentResource {
+    constructor(
+        name: string,
+        args: {
+            /**
+             * Alerts will be sent to this SNS topic
+             */
+            snsTopicArn: pulumi.Input<string>
+            /**
+             * The API Gateway to monitor
+             */
+            apiGateway: ApiGateway
+            /**
+             * Alert if the 5xx error rate exceeds this
+             * (default: 2 percent)
+             */
+            errorRate5xxPercent?: pulumi.Input<number>
+            /**
+             * a callback function that returns tags for
+             * each resource
+             */
+            getTags: (
+                name: string,
+            ) => {
+                [key: string]: pulumi.Input<string>
+            }
+        },
+        opts?: pulumi.ComponentResourceOptions,
+    ) {
+        super('wanews:apigateway/ErrorRate5xxAlarm', name, {}, opts)
+
+        const metricName = isHttpApi(args.apiGateway) ? '5xx' : '5XXError'
+        const dimensions = getDimensions(args.apiGateway)
+        const thresholdPercent = args.errorRate5xxPercent ?? 2
+        const threshold = pulumi
+            .output(thresholdPercent)
+            .apply((percent) => percent / 100)
+        const resourceName = `${name}-5xx-rate`
+
+        new aws.cloudwatch.MetricAlarm(
+            resourceName,
+            {
+                evaluationPeriods: 2,
+                datapointsToAlarm: 2,
+                comparisonOperator: 'GreaterThanOrEqualToThreshold',
+                namespace,
+                metricName,
+                dimensions,
+                threshold,
+                statistic: 'Average',
+                alarmDescription: pulumi.interpolate`threshold: 5xx rate >= ${thresholdPercent}%: ${
+                    args.apiGateway.name ?? args.apiGateway.id
```

Comment:
Can be simplified

Reply:
How so?

The name is best here, otherwise it just shows the ApiId, which isn't super-helpful.

Reply:
Wasn't the name needed for the REST api?

Reply:
The name here is used for the alarmDescription, I'll remove and show apiId instead.